File: src/main/java/de/bixilon/minosoft/data/mappings/blocks/Block.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.mappings.blocks;

import de.bixilon.minosoft.data.mappings.ModIdentifier;
import de.bixilon.minosoft.gui.rendering.chunk.models.BlockModel;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Objects;

public class Block extends ModIdentifier {
    private final BlockRotations rotation;
    private final HashSet<BlockProperties> properties;
    private BlockModel blockModel;

    public Block(String mod, String identifier, HashSet<BlockProperties> properties, BlockRotations rotation) {
        super(mod, identifier);
        this.properties = properties;
        this.rotation = rotation;
    }

    public Block(String mod, String identifier, HashSet<BlockProperties> properties) {
        super(mod, identifier);
        this.properties = properties;
        this.rotation = BlockRotations.NONE;
    }

    public Block(String fullIdentifier, BlockProperties... properties) {
        super(fullIdentifier);
        this.properties = new HashSet<>(Arrays.asList(properties));
        this.rotation = BlockRotations.NONE;
    }

    public Block(String mod, String identifier, BlockRotations rotation) {
        super(mod, identifier);
        this.properties = new HashSet<>();
        this.rotation = rotation;
    }

    public Block(String mod, String identifier) {
        super(mod, identifier);
        this.properties = new HashSet<>();
        this.rotation = BlockRotations.NONE;
    }

    public Block(String fullIdentifier) {
        super(fullIdentifier);
        this.properties = new HashSet<>();
        this.rotation = BlockRotations.NONE;
    }

    public BlockRotations getRotation() {
        return this.rotation;
    }

    public HashSet<BlockProperties> getProperties() {
        return this.properties;
    }

    public BlockModel getBlockModel() {
        return this.blockModel;
    }

    public void setBlockModel(BlockModel blockModel) {
        this.blockModel = blockModel;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.mod, this.identifier, this.properties, this.rotation);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (hashCode() != obj.hashCode()) {
            return false;
        }
        if (obj instanceof Block their) {
            return getIdentifier().equals(their.getIdentifier()) && getRotation() == their.getRotation() && getProperties().equals(their.getProperties()) && getMod().equals(their.getMod());
        }
        if (obj instanceof ModIdentifier identifier) {
            return super.equals(identifier);
        }
        return false;
    }

    public boolean bareEquals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj instanceof Block their) {
            if (!getMod().equals(their.getMod()) || !getIdentifier().equals(their.getIdentifier())) {
                return false;
            }
            if (their.getRotation() != BlockRotations.NONE) {
                if (their.getRotation() != getRotation()) {
                    return false;
                }
            }
            for (BlockProperties property : their.getProperties()) {
                if (!getProperties().contains(property)) {
                    return false;
                }
            }
            return true;
        }
        if (obj instanceof ModIdentifier identifier) {
            return super.equals(identifier);
        }
        return false;
    }

    @Override
    public String toString() {
        StringBuilder out = new StringBuilder();
        if (this.rotation != BlockRotations.NONE) {
            out.append(" (");
            out.append("rotation=");
            out.append(getRotation());
        }
        if (!this.properties.isEmpty()) {
            if (!out.isEmpty()) {
                out.append(", ");
            } else {
                out.append(" (");
            }
            out.append("properties=");
            out.append(this.properties);
        }
        if (!out.isEmpty()) {
            out.append(")");
        }
        return String.format("%s%s", getFullIdentifier(), out);
    }
}

File: src/main/java/de/bixilon/minosoft/data/mappings/blocks/Block.kt
/*
 * Minosoft
 * Copyright (C) 2021 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */
package de.bixilon.minosoft.data.mappings.blocks

import com.google.gson.JsonObject
import de.bixilon.minosoft.data.mappings.ResourceLocation
import de.bixilon.minosoft.data.mappings.items.Item
import de.bixilon.minosoft.data.mappings.registry.RegistryItem
import de.bixilon.minosoft.data.mappings.registry.ResourceLocationDeserializer
import de.bixilon.minosoft.data.mappings.versions.VersionMapping
import de.bixilon.minosoft.data.text.RGBColor
import de.bixilon.minosoft.gui.rendering.TintColorCalculator
import de.bixilon.minosoft.gui.rendering.chunk.models.renderable.BlockLikeRenderer

open class Block : RegistryItem {
    override val resourceLocation: ResourceLocation

    open val explosionResistance: Float
    open val tintColor: RGBColor?
    open val randomOffsetType: RandomOffsetTypes?
    open val tint: ResourceLocation?
    open val renderOverride: List<BlockLikeRenderer>? = null

    private val itemId: Int

    open lateinit var states: Set<BlockState>
        protected set
    open lateinit var defaultState: BlockState
        protected set
    open lateinit var item: Item
        protected set

    constructor(resourceLocation: ResourceLocation, mappings: VersionMapping, data: JsonObject) {
        resourceLocation.also { this.resourceLocation = it }
        (data["explosion_resistance"]?.asFloat ?: 0.0f).also { explosionResistance = it }
        data["tint_color"]?.asInt?.let { TintColorCalculator.getJsonColor(it) }.also { this@Block.tintColor = it }
        data["offset_type"]?.asString?.let { RandomOffsetTypes[it] }.also { randomOffsetType = it }
        itemId = data["item"]?.asInt ?: 0
        data["tint"]?.asString?.let { ResourceLocation(it) }.also { tint = it }
    }

    override fun postInit(versionMapping: VersionMapping) {
        item = versionMapping.itemRegistry.get(itemId)
    }

    override fun toString(): String {
        return resourceLocation.full
    }

    companion object : ResourceLocationDeserializer<Block> {
        override fun deserialize(mappings: VersionMapping?, resourceLocation: ResourceLocation, data: JsonObject): Block {
            check(mappings != null) { "VersionMapping is null!" }

            val block = when (data["class"].asString) {
                "FluidBlock" -> FluidBlock(resourceLocation, mappings, data)
                else -> Block(resourceLocation, mappings, data)
            }


            val states: MutableSet<BlockState> = mutableSetOf()
            for ((stateId, stateJson) in data["states"].asJsonObject.entrySet()) {
                check(stateJson is JsonObject) { "Not a state element!" }
                val state = BlockState.deserialize(block, mappings, stateJson, mappings.models)
                mappings.blockStateIdMap[stateId.toInt()] = state
                states.add(state)
            }

            block.states = states.toSet()
            block.defaultState = mappings.blockStateIdMap[data["default_state"].asInt]!!
            return block
        }
    }
}
